### PR TITLE
fix(behavior_path_planner): erase overlapping shifpoints

### PR DESCRIPTION
**Pull request by kyoichi-sugahara**
Signed-off-by: Kyoichi Sugahara [kyoichi.sugahara@tier4.jp](kyoichi.sugahara@tier4.jp)

## Description
In the sideshift module, there was no function to erase existing shift points when a new offset value is given.
As a result, unstable routes were generated as different sets of shift points intersected within a certain section.
With this change, if a newly added shift point overlaps a straight line created from existing shift points, the existing shift points are erased.

##Pre-review checklist for the PR author
The PR author **must** check the checkboxes below when creating the PR.

 

- [x]  I've confirmed the [contribution guidelines](https://autowarefoundation.github.io/autoware-documentation/main/contributing/).
- [x]  The PR follows the [pull request guidelines](https://autowarefoundation.github.io/autoware-documentation/main/contributing/pull-request-guidelines/).

## In-review checklist for the PR reviewers
The PR reviewers **must** check the checkboxes below before approval.
- [ ] The PR follows the [pull request guidelines](https://autowarefoundation.github.io/autoware-documentation/main/contributing/pull-request-guidelines/).

## Post-review checklist for the PR author
The PR author must check the checkboxes below before merging.

- [ ] There are no open discussions or they are tracked via tickets.

After all checkboxes are checked, anyone who has write access can merge the PR.

### CODE REVIEW

**Comment by zulfaqar-azmi-t4** [comment]
The result looks promising.
Tested the reset with `shift_request_time_limit: 0.5`

https://user-images.githubusercontent.com/93502286/177909758-05079f9d-072c-41e0-b5ff-e64fa7d397fa.mp4

Tested the reset with `shift_request_time_limit: 0.2`

https://user-images.githubusercontent.com/93502286/177909883-e4b8eae0-0918-44d5-9a7c-4147f67f4cc4.mp4

File: planning/behavior_path_planner/src/scene_module/side_shift/side_shift_module.cpp [hunk 1]
```diff
@@ -223,13 +228,39 @@ bool SideShiftModule::addShiftPoint()
...
+  const auto new_sp_longitudinal_to_shift_start = calcLongitudinal_to_shift_start(new_sp);
+  const auto new_sp_longitudinal_to_shift_end = calcLongitudinal_to_shift_end(new_sp);
+
+  const auto remove_overlap_iter = std::remove_if(
+    shift_points.begin(), shift_points.end(),
+    [this, calcLongitudinal_to_shift_start, calcLongitudinal_to_shift_end,
+     new_sp_longitudinal_to_shift_start, new_sp_longitudinal_to_shift_end](const ShiftPoint & sp) {
+      const auto old_sp_longitudinal_to_shift_start = calcLongitudinal_to_shift_start(sp);
+      const auto old_sp_longitudinal_to_shift_end = calcLongitudinal_to_shift_end(sp);
+      bool sp_overlap_front =
```
Comment (taikitanaka3) [inline planning/behavior_path_planner/src/scene_module/side_shift/side_shift_module.cpp:244]: @kyoichi-sugahara 
can you add const here?

Comment (taikitanaka3) [inline planning/behavior_path_planner/src/scene_module/side_shift/side_shift_module.cpp:244]:
```suggestion
      const bool sp_overlap_front =
```

**Comment by taikitanaka3** [comment]
LGTM
It works fine with the nominal test case
https://user-images.githubusercontent.com/65527974/177916339-74d3a8a0-e6e9-4fda-abb7-8a18f80fad4b.mp4

**Review by taikitanaka3 — APPROVED** [review]
LGTM

TODO: consider more smooth new shift point after deleting old shift point

**Pull request merged**